contracts/MarketMint.sol

```diff
@@ -130,91 +128,124 @@ contract Mint is ReentrancyGuard, Pausable {
-  <~~~*/
+    /*~~~> 
+      uint _redeemAmount: amount needed to redeem for creating new NFTs;
+      address _contract: address for the token;
+    <~~~*/
   /// @return Bool
-  function setNewRedemption(uint amount, address _toke) public hasAdmin returns(bool){
+  function setNewRedemption(uint _redeemAmount, address _contract) external hasAdmin returns(bool){
     _redemptionERC20.increment();
     uint id = _redemptionERC20.current();
-    _idToRedemption[id] = RedemptionToken(amount, _toke);
+    _idToRedemption[id] = RedemptionToken(_redeemAmount, _contract);
     return true;
   }
 
   /// @notice
-  /*~~~>
-  Public interaction function for creating new NFTs by exchanging Tokens
-  <~~~*/
+  //*~~~> For updating the total NFT array with new tokenIds
+    ///@dev Can only be incremented!
+    //*~~~> uint _tokenIds: token Ids to add to the array
+  function setNftTokenIds(uint _amount) external hasDevAdmin returns(bool){
+    for (uint i; i<_amount; i++){
+      totalNfts = totalNfts+1;
+      availableNfts.push(totalNfts);
+    }
+    nftsRemaining = availableNfts.length;
+    return true;
+  }
+
+  /// @notice
+    /*~~~>
+      Public interaction function for redeeming new NFTs by exchanging Tokens
+    <~~~*/
   /// @dev
-  /*~~~>
-    uint id: index to the redemption token struct;
-    uint amount: amount needed to redeem;
-    address to: address to mint the NFT to;
-  <~~~*/
+    /*~~~>
+      uint id: index to the redemption token struct;e
+    <~~~*/
   /// @return Bool
-  function redeemForNft(uint id, uint amount, address to) public whenNotPaused returns(bool){
+  function redeemForNft(uint id) external whenNotPaused returns(bool){
 
     address rewardsAddress =  IRoleProvider(roleAdd).fetchAddress(REWARDS);
     address nftAddress = IRoleProvider(roleAdd).fetchAddress(NFTADD);
 
+    /// Bring in token Interface
     RedemptionToken memory token = _idToRedemption[id];
-    require(amount >= token.redeemAmount);
-
     IERC20 tokenContract = IERC20(token.contractAddress);
+    /// Check allowance
     uint256 allowance = tokenContract.allowance(msg.sender, address(this));
     require(allowance >= token.redeemAmount, "Check the token allowance");
+    /// Execute transfer
     tokenContract.transferFrom(msg.sender, rewardsAddress, token.redeemAmount);
     
-    MarketNFT(nftAddress).safeMint(to);
-    _nftsCreated.increment();
-    uint256 nftId = _nftsCreated.current();
-    _idToNft[nftId] = NFT(nftId, msg.sender);
-    IRewardsController(rewardsAddress).depositERC20Rewards(token.redeemAmount, token.contractAddress);
-    IRewardsController(rewardsAddress).createNftHodler(nftId);
+    ///Keeping track of total claimed
+    _nftsRedeemed.increment();
+    uint256 nftId = _nftsRedeemed.current();
+    /// Using the new ID as a nonce to generate a random tokenId from the available NFTs
+    uint tokenId = getQuasiRandom(nftId);
+    IERC721(nftAddress).safeTransferFrom(address(this), msg.sender, tokenId);
```

Comment:
How does the SafeTransfer line work? None of the tokens are minted, am I missing something?

Reply:
They are being preminted and sent to this contract.